src/mysqlsh_helpers.py

```diff
@@ -311,6 +367,200 @@ def add_instance_to_cluster(self, instance_address, instance_unit_label) -> None
+        )
+
+        try:
+            logger.debug(
+                f"Confirming instance {instance_address}/{instance_unit_label} configuration for InnoDB"
+            )
+
+            output = self._run_mysqlsh_script("\n".join(commands))
+            return "INSTANCE_CONFIGURED" in output
+        except subprocess.CalledProcessError as e:
+            # confirmation can fail if the clusteradmin user does not yet exist on the instance
+            logger.warning(
+                f"Failed to confirm instance configuration for {instance_address} with error {e.stderr}",
+                exc_info=e,
+            )
+            return False
+
+    @retry(
+        retry=retry_if_exception_type(MySQLRemoveInstanceError),
+        stop=stop_after_attempt(15),
+        reraise=True,
+        wait=wait_random(min=4, max=30),
+    )
+    def remove_instance(self, unit_label: str) -> None:
+        """Remove instance from the cluster.
+
+        This method is called from each unit being torn down, thus we must obtain
+        locks on the cluster primary. There is a retry mechanism for any issues
+        obtaining the lock, removing instances/dissolving the cluster, or releasing
+        the lock.
+
+        Args:
+            unit_label: The label for this unit's instance (to be torn down)
+        """
+        try:
+            # Get the cluster primary's address to direct lock acquisition request to.
+            primary_address = self._get_cluster_primary_address()
+            if not primary_address:
+                raise MySQLRemoveInstanceError("Unable to retrieve the cluster primary's address")
+
+            # Attempt to acquire a lock on the primary instance
+            logger.debug(
+                f"Attempting to acquire lock to remove unit {self.instance_address}/{unit_label}"
+            )
+            acquire_lock_commands = (
+                f"shell.connect('{self.cluster_admin_user}:{self.cluster_admin_password}@{primary_address}')",
+                f"session.run_sql(\"UPDATE mysql.juju_units_operations SET executor='{unit_label}', status='in-progress' WHERE task='unit-teardown' AND executor='';\")",
+                f"acquired_lock = session.run_sql(\"SELECT count(*) FROM mysql.juju_units_operations WHERE task='unit-teardown' AND executor='{unit_label}';\").fetch_one()[0]",
+                "print(f'<ACQUIRED_LOCK>{acquired_lock}</ACQUIRED_LOCK>')",
+            )
+
+            output = self._run_mysqlsh_script("\n".join(acquire_lock_commands))
+            matches = re.search(r"<ACQUIRED_LOCK>(\d)</ACQUIRED_LOCK>", output)
+            if not matches:
+                raise MySQLRemoveInstanceError("Unable to acquire lock to remove unit")
+
+            # If unable to acquire the lock, raise an error to retry
+            acquired_lock = int(matches.group(1))
+            if not acquired_lock:
+                raise MySQLRemoveInstanceError("Did not acquire lock to remove unit")
```

Comment:
This method's pretty big - would it makes sense to refactor out the lock logic to individual `get_lock` and `remove_lock` methods?